// @midwayjs/swagger@3.20.13/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DECORATORS = exports.DECORATORS_METHOD_METADATA = exports.DECORATORS_CLASS_METADATA = exports.DECORATORS_PREFIX = void 0;
exports.DECORATORS_PREFIX = 'swagger';
exports.DECORATORS_CLASS_METADATA = 'swagger:class_metadata';
exports.DECORATORS_METHOD_METADATA = 'swagger:method_metadata';
exports.DECORATORS = {
    API_OPERATION: `${exports.DECORATORS_PREFIX}/apiOperation`,
    API_RESPONSE: `${exports.DECORATORS_PREFIX}/apiResponse`,
    API_TAGS: `${exports.DECORATORS_PREFIX}/apiUseTags`,
    API_PARAMETERS: `${exports.DECORATORS_PREFIX}/apiParameters`,
    API_HEADERS: `${exports.DECORATORS_PREFIX}/apiHeaders`,
    API_MODEL_PROPERTIES: `${exports.DECORATORS_PREFIX}/apiModelProperties`,
    API_EXTRA_MODEL: `${exports.DECORATORS_PREFIX}/apiExtraModel`,
    API_MODEL_PROPERTIES_ARRAY: `${exports.DECORATORS_PREFIX}/apiModelPropertiesArray`,
    API_SECURITY: `${exports.DECORATORS_PREFIX}/apiSecurity`,
    API_EXCLUDE_SECURITY: `${exports.DECORATORS_PREFIX}/apiExcludeSecurity`,
    API_EXCLUDE_ENDPOINT: `${exports.DECORATORS_PREFIX}/apiExcludeEndpoint`,
    API_EXCLUDE_CONTROLLER: `${exports.DECORATORS_PREFIX}/apiExcludeController`,
    API_EXTENSION: `${exports.DECORATORS_PREFIX}/apiExtension`,
};
