// @midwayjs/swagger@3.20.13/dist/decorators/api-property.decorator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiResponseProperty = exports.ApiPropertyOptional = exports.createApiPropertyDecorator = exports.ApiProperty = void 0;
const constants_1 = require("../constants");
const enum_utils_1 = require("../common/enum.utils");
const helpers_1 = require("./helpers");
const isEnumArray = (obj) => obj.isArray && !!obj.enum;
function ApiProperty(options = {}) {
    return createApiPropertyDecorator(options);
}
exports.ApiProperty = ApiProperty;
function createApiPropertyDecorator(options = {}) {
    const [type, isArray] = (0, helpers_1.getTypeIsArrayTuple)(options.type, options.isArray);
    options = {
        ...options,
        type,
    };
    if (isEnumArray(options)) {
        options.type = 'array';
        const enumValues = (0, enum_utils_1.getEnumValues)(options.enum);
        options.items = {
            type: (0, enum_utils_1.getEnumType)(enumValues),
            enum: enumValues,
        };
        delete options.enum;
    }
    else if (options.enum) {
        const enumValues = (0, enum_utils_1.getEnumValues)(options.enum);
        options.enum = enumValues;
        options.type = (0, enum_utils_1.getEnumType)(enumValues);
    }
    if (isArray) {
        options.type = 'array';
        options.items = {
            type: type,
        };
    }
    return (0, helpers_1.createPropertyDecorator)(constants_1.DECORATORS.API_MODEL_PROPERTIES, options);
}
exports.createApiPropertyDecorator = createApiPropertyDecorator;
function ApiPropertyOptional(options = {}) {
    return ApiProperty({
        ...options,
        required: false,
    });
}
exports.ApiPropertyOptional = ApiPropertyOptional;
function ApiResponseProperty(options = {}) {
    return ApiProperty({
        readOnly: true,
        ...options,
    });
}
exports.ApiResponseProperty = ApiResponseProperty;
