// @midwayjs/swagger@3.20.13/dist/decorators/api-header.decorator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiHeaders = exports.ApiHeader = void 0;
const constants_1 = require("../constants");
const enum_utils_1 = require("../common/enum.utils");
const helpers_1 = require("./helpers");
const defaultHeaderOptions = {
    name: '',
};
function ApiHeader(options) {
    const param = {
        name: !options.name ? defaultHeaderOptions.name : options.name,
        in: 'header',
        description: options.description,
        required: options.required,
        schema: {
            ...(options.schema || {}),
            type: 'string',
        },
    };
    if (options.enum) {
        const enumValues = (0, enum_utils_1.getEnumValues)(options.enum);
        param.schema = {
            enum: enumValues,
            type: (0, enum_utils_1.getEnumType)(enumValues),
        };
    }
    return (target, key, descriptor) => {
        if (descriptor) {
            return (0, helpers_1.createMixedDecorator)(constants_1.DECORATORS.API_HEADERS, param)(target, key, descriptor);
        }
        return (0, helpers_1.createMixedDecorator)(constants_1.DECORATORS.API_HEADERS, param)(target, undefined, undefined);
    };
}
exports.ApiHeader = ApiHeader;
const ApiHeaders = (headers) => {
    return (target, key, descriptor) => {
        headers.forEach(options => ApiHeader(options)(target, key, descriptor));
    };
};
exports.ApiHeaders = ApiHeaders;
