// @midwayjs/swagger@3.20.13/dist/decorators/api-body.decorator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiBody = exports.BodyContentType = void 0;
const enum_utils_1 = require("../common/enum.utils");
const helpers_1 = require("./helpers");
var BodyContentType;
(function (BodyContentType) {
    BodyContentType["FormUrlEncoded"] = "application/x-www-form-urlencoded";
    BodyContentType["JSON"] = "application/json";
    BodyContentType["Multipart"] = "multipart/form-data";
    BodyContentType["MultipartMixed"] = "multipart/mixed";
    BodyContentType["OctetStream"] = "application/octet-stream";
})(BodyContentType = exports.BodyContentType || (exports.BodyContentType = {}));
const defaultBodyMetadata = {
    type: String,
    required: true,
};
function ApiBody(options) {
    const [type, isArray] = (0, helpers_1.getTypeIsArrayTuple)(options.type, options.isArray);
    const param = {
        in: 'body',
        ...options,
        type,
        isArray,
    };
    if ((0, enum_utils_1.isEnumArray)(options)) {
        (0, enum_utils_1.addEnumArraySchema)(param, options);
    }
    else if ((0, enum_utils_1.isEnumDefined)(options)) {
        (0, enum_utils_1.addEnumSchema)(param, options);
    }
    return (0, helpers_1.createParamDecorator)(param, defaultBodyMetadata);
}
exports.ApiBody = ApiBody;
